File: python/day-03/src/day_03/part_one.py
from dataclasses import dataclass
from typing import List, Tuple


@dataclass
class Point:
    x: int
    y: int
    val: str

    def vector_in_aoe(self, vec: "Vector") -> bool:
        if (self.y - 1) > vec.y or (self.y + 1) < vec.y:
            return False

        if (self.x - 1) > vec.x_max or (self.x + 1) < vec.x_min:
            return False

        return True


@dataclass
class Vector:
    x_min: int
    x_max: int
    y: int
    _val: str

    @classmethod
    def from_point(cls, point: Point):
        return cls(x_min=point.x, x_max=point.x, y=point.y, _val=point.val)

    def add_point(self, point: Point):
        assert point.y == self.y

        if point.x == self.x_min - 1:
            self.x_min = point.x
            self._val = point.val + self._val
        elif point.x == self.x_max + 1:
            self.x_max = point.x
            self._val += point.val

    @property
    def val(self):
        return int(self._val)


def read_input(fp: str) -> List[str]:
    with open(fp, "r") as file:
        input = [line.strip() for line in file.readlines()]

    return input


def part_one(input: List[str]) -> int:
    symbols, numbers = find_symbols_and_numbers(input)

    engine_schematic_sum = 0
    for number in numbers:
        for symbol in symbols:
            if symbol.vector_in_aoe(number):
                engine_schematic_sum += number.val
                break

    return engine_schematic_sum


def find_symbols_and_numbers(
    input: List[List[str]],
) -> Tuple[List[Point], List[Vector]]:
    symbol_coordinates = []
    number_vectors = []

    for idx_y, line in enumerate(input):
        current_vector = None
        for idx_x, col in enumerate(line):
            if not col.isnumeric():
                if current_vector:
                    number_vectors.append(current_vector)
                    current_vector = None

                if col != ".":
                    point = Point(idx_x, idx_y, col)
                    symbol_coordinates.append(point)

            if col.isnumeric():
                point = Point(idx_x, idx_y, col)
                if not current_vector:
                    current_vector = Vector.from_point(point)
                else:
                    current_vector.add_point(point)

        if current_vector:
            number_vectors.append(current_vector)
            current_vector = None

    return (symbol_coordinates, number_vectors)


def main():
    input = read_input("input.txt")
    output = part_one(input)
    print(output)


if __name__ == "__main__":
    main()

File: rust/day-03/src/bin/part_one.rs
use std::cmp::min;

#[derive(Debug)]
struct Point {
    x: u32,
    y: u32,
    val: String,
}

impl Point {
    fn vector_in_aoe(&self, vector: &Vector) -> bool {
        if self.y == 0 && (self.y + 1) < vector.y {
            return false;
        }

        if (self.y - 1) > vector.y || (self.y + 1) < vector.y {
            return false;
        }

        if self.x == 0 && (self.x + 1) < vector.x_min {
            return false;
        }

        if (self.x - 1) > vector.x_max || (self.x + 1) < vector.x_min {
            return false;
        }

        true
    }
}

#[derive(Debug)]
struct Vector {
    x_min: u32,
    x_max: u32,
    y: u32,
    val: u32,
}

impl Vector {
    fn from_point(point: Point) -> Self {
        Self {
            x_min: point.x,
            x_max: point.x,
            y: point.y,
            val: point.val.parse::<u32>().unwrap(),
        }
    }

    fn add_point(&mut self, point: Point) {
        assert_eq!(self.y, point.y);

        if min(0, self.x_min.wrapping_sub(1)) == point.x {
            self.x_min = point.x;
            self.val = format!("{0}{1}", point.val, self.val)
                .parse::<u32>()
                .unwrap();
        } else if self.x_max + 1 == point.x {
            self.x_max = point.x;
            self.val = format!("{0}{1}", self.val, point.val)
                .parse::<u32>()
                .unwrap();
        }
    }

    fn valid_point(&self, point: &Point) -> bool {
        if self.y != point.y {
            return false;
        }

        if min(0, self.x_min.wrapping_sub(1)) != point.x {
            if self.x_max + 1 != point.x {
                return false;
            }
        }

        true
    }
}

fn main() {
    let input = include_str!("./input.txt");
    let output = part_one(&input);
    dbg!(output);
}

fn part_one(input: &str) -> u32 {
    let symbols = find_symbols(input);
    let numbers = find_numbers(input);

    let mut engine_schematic_sum = 0;
    for number in &numbers {
        for symbol in &symbols {
            if symbol.vector_in_aoe(number) {
                engine_schematic_sum += number.val;
                break;
            }
        }
    }

    engine_schematic_sum
}

fn find_symbols(input: &str) -> Vec<Point> {
    let mut symbols = vec![];
    for (idx_y, line) in input.lines().enumerate() {
        for (idx_x, char) in line.chars().enumerate() {
            if !char.is_ascii_digit() && char != '.' {
                let symbol = Point {
                    x: idx_x as u32,
                    y: idx_y as u32,
                    val: String::from(char),
                };
                symbols.push(symbol);
            }
        }
    }

    symbols
}

fn find_numbers(input: &str) -> Vec<Vector> {
    let mut numbers: Vec<Vector> = vec![];

    for (idx_y, line) in input.lines().enumerate() {
        for (idx_x, char) in line
            .chars()
            .enumerate()
            .filter(|(_, char)| char.is_ascii_digit())
        {
            let point = Point {
                x: idx_x as u32,
                y: idx_y as u32,
                val: String::from(char),
            };

            match numbers.last_mut() {
                Some(vec) => {
                    if vec.valid_point(&point) {
                        vec.add_point(point)
                    } else {
                        numbers.push(Vector::from_point(point))
                    }
                }
                None => numbers.push(Vector::from_point(point)),
            }
        }
    }

    numbers
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        let result = part_one(
            "467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..",
        );
        assert_eq!(result, 4361)
    }
}
